pandas/tests/sparse/test_frame.py

```diff
@@ -1267,6 +1268,33 @@ def test_from_scipy_correct_ordering(spmatrix):
     tm.assert_frame_equal(sdf.to_dense(), expected.to_dense())
 
 
+def test_from_scipy_object_fillna(spmatrix):
+    # GH 16112
+    tm.skip_if_no_package('scipy', max_version='0.19.0')
+
+    arr = np.eye(3)
+    arr[1:, 0] = np.nan
+
+    try:
+        spm = spmatrix(arr)
+        assert spm.dtype == arr.dtype
+    except (TypeError, AssertionError):
+        # If conversion to sparse fails for this spmatrix type and arr.dtype,
+        # then the combination is not currently supported in NumPy, so we
+        # can just skip testing it thoroughly
+        return
+
+    sdf = pd.SparseDataFrame(spm).fillna(-1.0)
+
+    # Returning frame should fill all nan values with -1.0
+    expected = pd.SparseDataFrame({0: {0: 1.0, 1: np.nan, 2: np.nan},
+                                   1: {0: np.nan, 1: 1.0, 2: np.nan},
+                                   2: {0: np.nan, 1: np.nan, 2: 1.0}}
+                                  ).fillna(-1.0)
+
+    tm.assert_frame_equal(sdf.to_dense(), expected.to_dense())
```

Comment:
Replacing `tm.assert_frame_equal(sdf.to_dense(), expected.to_dense())` in the above code (line 1295) with `tm.assert_sp_frame_equal(sdf, expected)` results in the following error:

```python
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas/util/testing.py:1583: in assert_sp_frame_equal
    check_dtype=check_dtype)
pandas/util/testing.py:1538: in assert_sp_series_equal
    assert_sp_array_equal(left.block.values, right.block.values)
pandas/util/testing.py:1493: in assert_sp_array_equal
    check_dtype=check_dtype)
pandas/util/testing.py:1176: in assert_numpy_array_equal
    _raise(left, right, err_msg)
pandas/util/testing.py:1159: in _raise
    .format(obj), left.shape, right.shape)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
obj = 'numpy array', message = 'numpy array shapes are different', left = (3,), right = (1,)
diff = None

    def raise_assert_detail(obj, message, left, right, diff=None):
        if isinstance(left, np.ndarray):
            left = pprint_thing(left)
        if isinstance(right, np.ndarray):
            right = pprint_thing(right)

        msg = """{0} are different

    {1}
    [left]:  {2}
    [right]: {3}""".format(obj, message, left, right)

        if diff is not None:
            msg = msg + "\n[diff]: {diff}".format(diff=diff)

>       raise AssertionError(msg)
E       AssertionError: numpy array are different
E
E       numpy array shapes are different
E       [left]:  (3,)
E       [right]: (1,)

pandas/util/testing.py:1109: AssertionError
```